### PR TITLE
Fail when a template does not exist

### PR DESCRIPTION
**Pull request by jan-cerny**
#### Description:
The current code tries to generate content from any template it finds.
However, that means when it can't find the template it skips it and
continues.  If an existing template goes missing, or is not found by any
reason, it will quietly continue, and  the rule evaluation result in
notchecked without evident reason. 

As suggested by @matejak we will use Python decorators to declare which languages are available for each template. That is better than having the manually created dictionary.

#### Rationale:

In #4809 @yuumasato pointed out problems caused by implicit behavior - see discussion there.

### CODE REVIEW

File: ssg/templates.py [hunk 1]
```diff
@@ -238,44 +238,108 @@ def timer_enabled(data, lang):
...
-    "sebool": None,
-    "sebool_var": None,
-    "service_disabled": service_disabled,
-    "service_enabled": service_enabled,
-    "sshd_lineinfile": sshd_lineinfile,
-    "sysctl": sysctl,
-    "timer_enabled": timer_enabled,
-}
+    "accounts_password": (
+        ["ansible", "bash", "oval"],
```
[the remark this answers is not in the document]
Comment (jan-cerny) [reply ssg/templates.py]: @matejak check!

**Comment by jan-cerny** [comment]
@matejak @ggbecker @jan-cerny @matusmarhefka @mildas @yuumasato @vojtapolasek @dahaic Any volunteers for review?

**Comment by jan-cerny** [comment]
I have improved the documentation and rebased it on master.

**Comment by ggbecker** [comment]
I like very much the idea of using decorators, but I'm feeling that the way it's implemented generates lots of duplicated code, for example:

`@template(["ansible", "bash", "oval"])`, those strings are basically repeated in every templated function, apart from one or another which contains puppets or doesn't contain some `language`.

I would like to propose an enhanced solution of current implementation. Instead of specifying the available languages for a given template, why not let the build system identify automatically which of the `languages` are available for this given template. This would imply that the build system needs to look into the `template_<LANG>_<NAME>` files and decided whether one language must be included or not. And I also would propose to get rid of `template_<LANG>_<NAME>` files by having them split into directories such as: `shared/templates/ansible/`, `shared/templates/bash/`, `shared/templates/oval/` and so on. This way would be much cleaner how to find the respective templated files.

I'm not blocking this PR to be merged, it has already improved a lot from the previous state but I still think there is a lot to improve.

File: ssg/templates.py [hunk 2]
```diff
@@ -16,15 +16,28 @@
     "puppet": ".pp"
 }
 
+templates = dict()
+
+
+def template(langs):
+    def decorator_template(func):
+        func.langs = langs
+        templates[func.__name__] = func
```
Comment (matejak) [inline ssg/templates.py]: This is a great improvement!

**Comment by matejak** [comment]
Reacting to @ggbecker's suggestion:
It is a matter of taste what to count as the source of truth. Approaches towards deducing what templated content we support is are these:

-  Take a look at available template files. This is minimalistic and good enough to use whatever templates are available, and to produce valid datastreams.
- Declare the templated content in the source code. Then, if a template file is missing, it will raise an error, and if it is present, but the association is not declared, it won't be used.

In other words, the second approach adds cross-checking/intermediate level of control. Let's go for it as it is implemented already - if we find out that it is too complex, we can always simplify it.

**Comment by matejak** [comment]
Decorators in template definitions open a path towards more robust, while easy to understand templating setup. This gets rid of the `templates` mapping. Well done!
